### ISSUE TITLE
watchdog timeout on close

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#647]
I've noticed that upgrading tailscaled deb packages has been slow sometimes.

Seems like the shutdown part is blocking. Relevant stacks:  (from cff737786ed9907fa3c9da9d6001b5d6c8f1a315)

```
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: 6 @ 0x437230 0x448060 0x44804b 0x447dc7 0x47338c 0x855956 0x85575b 0x4671f1
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x447dc6        sync.runtime_SemacquireMutex+0x46                                                runtime/sema.go:71
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x47338b        sync.(*Mutex).lockSlow+0xfb                                                        sync/mutex.go:138
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x855955        sync.(*Mutex).Lock+0x375                                                        sync/mutex.go:81
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x85575a        github.com/tailscale/wireguard-go/device.(*Peer).RoutineSequentialSender+0x17a        github.com/tailscale/wireguard-go@v0.0.0/device/send.go:620
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: 1 @ 0x437230 0x448060 0x44804b 0x447cb2 0x474e04 0x84fae0 0x85ada9 0x849c5f 0x849f59 0x90ecad 0x912774 0x912707 0x4671f1
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x447cb1        sync.runtime_Semacquire+0x41                                                        runtime/sema.go:56
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x474e03        sync.(*WaitGroup).Wait+0x63                                                        sync/waitgroup.go:130
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x84fadf        github.com/tailscale/wireguard-go/device.(*Peer).Stop+0x12f                        github.com/tailscale/wireguard-go@v0.0.0/device/peer.go:312
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x85ada8        github.com/tailscale/wireguard-go/device.(*Device).RemoveAllPeers.func1+0x48        github.com/tailscale/wireguard-go@v0.0.0/device/device.go:419
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x849c5e        github.com/tailscale/wireguard-go/device.(*Device).RemoveAllPeers+0x2ce                github.com/tailscale/wireguard-go@v0.0.0/device/device.go:430
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x849f58        github.com/tailscale/wireguard-go/device.(*Device).Close+0x138                        github.com/tailscale/wireguard-go@v0.0.0/device/device.go:480
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x90ecac        tailscale.com/wgengine.(*userspaceEngine).Close+0x35c                                tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/userspace.go:1064
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x912773        tailscale.com/wgengine.(*watchdogEngine).watchdog.func1+0x23                        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/watchdog.go:68
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x912706        tailscale.com/wgengine.(*watchdogEngine).watchdogErr.func1+0x26                        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/watchdog.go:50
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: 1 @ 0x755b65 0x755980 0x75271a 0x91052a 0x910749 0x91118b 0x916147 0x957ced 0x963c1a 0x9634f3 0x436e6a 0x4671f1
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x755b64        runtime/pprof.writeRuntimeProfile+0x94                                runtime/pprof/pprof.go:694
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x75597f        runtime/pprof.writeGoroutine+0x9f                                runtime/pprof/pprof.go:656
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x752719        runtime/pprof.(*Profile).WriteTo+0x3d9                                runtime/pprof/pprof.go:329
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x910529        tailscale.com/wgengine.(*watchdogEngine).watchdogErr+0x1e9        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/watchdog.go:59
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x910748        tailscale.com/wgengine.(*watchdogEngine).watchdog+0x78                tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/watchdog.go:67
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x91118a        tailscale.com/wgengine.(*watchdogEngine).Close+0x9a                tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/watchdog.go:113
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x916146        tailscale.com/ipn.(*LocalBackend).Shutdown+0x86                        tailscale.com@v0.0.0-00010101000000-000000000000/ipn/local.go:128
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x957cec        tailscale.com/ipn/ipnserver.Run+0xf5c                                tailscale.com@v0.0.0-00010101000000-000000000000/ipn/ipnserver/server.go:274
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x963c19        main.run+0x5b9                                                        tailscale.com@v0.0.0-00010101000000-000000000000/cmd/tailscaled/tailscaled.go:177
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x9634f2        main.main+0x4e2                                                        tailscale.com@v0.0.0-00010101000000-000000000000/cmd/tailscaled/tailscaled.go:106
Aug 06 16:35:44 hello.ipn.dev tailscaled[4687]: #        0x436e69        runtime.main+0x1f9                                                runtime/proc.go:203
Aug 06 16:35:45 hello.ipn.dev tailscaled[4687]: wgengine: watchdog timeout on Close
```

/cc @josharian @dshynkev @danderson

**Comment by DentonGentry** [comment]
Barring further comments, we'll expect to close this during the next bug scrub.

**Comment by DentonGentry** [comment]
To my knowledge, we haven't noticed this again since the bug was filed over a year ago.
Barring objection, we'll expect to close this during the next bug scrub.

**Comment by DentonGentry** [comment]
ETIMEDOUT.

**Comment by bradfitz** [comment]
Actually I think we fixed this the other day and was in 1.18.2. The systemd resolved restart thing.

**Comment by DentonGentry** [comment]
Fixed by https://github.com/tailscale/tailscale/commit/2d6404609d9d07557379be15e0772e98afc966d3
